sound/soc/sof/intel/hda.c

```diff
@@ -251,15 +261,43 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	}
 
 	/* codec detection */
-	if (!bus->codec_mask)
+	if (!bus->codec_mask) {
 		dev_info(bus->dev, "no hda codecs found!\n");
-	else
-		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
+	} else {
+		dev_info(bus->dev, "hda codecs found, mask %lx\n",
+			 bus->codec_mask);
+
+		for (i = 0; i < HDA_MAX_CODECS; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/*
+		 * If there are HDMI codec and other HDA codecs, hda machine
+		 * driver is used. For I2S codec + HDMI codec, there is only one
+		 * HDMI codec. If only one or none of HDMI codec is found,
+		 * nocodec machine driver would be used if no I2S codec machine
+		 * is used.
```

Comment:
This is a good start but we can make it more general to handle multiple situations:
0. codec_num = 0 -> use nocodec if enabled in the build
1. codec_num = 1 -> use nocodec if enabled in the build (for validation)
2. codec_num = 1 -> use HDMI (for user interaction e.g. on a headless device like Up2)
3. codec_num = 2 -> use HDMI + HDAudio

Also for HDMI you need to verify that bit2 is set, it's not just the number of codecs that matters.
Also you would need to set the HDAudio machine driver only if there is no ACPI ID (or machine already set). 

Makes sense?